Endpoint to populate Status Box [Issue #142]

mahieakhtar opened the pull request:
resolves #142 
Status label is not part of this PR.

sowmya1408 commented on src/server/api/routes/status-box.router.js:
```diff
@@ -22,4 +22,20 @@ router.get(
 
+// ENDPOINT: /api/status-box/daysTilHarvest/:batchId :GET to get days till harvest date of a batch with batchId
+router.get('/days-to-harvest/:batchId', (req, res, next) => {
+  statusBoxController
```
the middleware checkIfAuthenticated is missing in this endpoint, may be you can refer to the : firebase-middlewares.md document in our project to on how to use it on end point :)

sowmya1408 commented on src/server/api/routes/status-box.router.js:
```diff
@@ -22,4 +22,20 @@ router.get(
+
+// ENDPOINT: /api/status-box/productionDate/:batchId :GET to get production date
+router.get('/production-date/:batchId', (req, res, next) => {
+  statusBoxController
```
Even here the middleware is missing, just refer to above comments :)

sowmya1408 commented on src/server/config/swagger.json:
```diff
@@ -97,42 +97,72 @@
+            "name": "batchId",
+            "in": "path",
+            "type": "integer",
+            "required": true
```
There should also be a firebase object, which give you the option to enter the token in swagger is nice to add, as it would know that you are authenticated.
`{
            "name": "authorization",
            "in": "header",
            "description": "Firebase token",
            "required": true,
            "type": "string",
            "default": "Bearer ENTER_FIREBASE_TOKEN"
          }`

sowmya1408 commented on src/server/api/routes/status-box.router.js:
```diff
@@ -0,0 +1,25 @@
+// ENDPOINT: /api/status-box/current-stage/:batchId :GET to get current stage message of a batch with batchId
+router.get(
+  '/current-stage/:batchId',
+  checkIfAuthenticated,
```
Good to see the endpoint is protected using middleware :)

sowmya1408 commented on src/server/config/swagger.json:
```diff
@@ -202,6 +202,48 @@
+            "required": true,
+            "type": "string",
+            "default": "Bearer ENTER_FIREBASE_TOKEN"
+          },
```
👍

sowmya1408 commented on src/server/api/routes/days-to-end-batch.router.js:
```diff
@@ -0,0 +1,17 @@
+const dayLeftController = require('../controllers/days-to-end-batch.controllers');
+
+// ENDPOINT: /api/day-left/:batchId :GET to get Days left to end the batch
+router.get('/:batchId', (req, res, next) => {
```
its good to protect the endpoint with a middleware.Could you refer to firebase-middleware.md file in our project on how to use middleware on endpoint

sowmya1408 commented on src/server/config/swagger.json:
```diff
@@ -9,6 +9,40 @@
+            "in": "path",
+            "type": "integer",
+            "required": true
+          }
```
An object which allows to input the firebase token should also be added to swagger json.something like this:
`{
            "name": "authorization",
            "in": "header",
            "description": "Firebase token",
            "required": true,
            "type": "string",
            "default": "Bearer ENTER_FIREBASE_TOKEN"
          },`

mahieakhtar commented:
Please update fk_crops_id to fk_crop_id in your knex migration and controller if it's being used because fk_crops_id has been updated to fk_crop_id to make it consistent and to avoid confusion on migrations schema and DB diagram. So please replace fk_crops_id to fk_crop_id if your application is crashing and you want to save time.

mahieakhtar commented on src/server/api/controllers/days-to-end-batch.controllers.js:
```diff
@@ -0,0 +1,31 @@
+
+    const totalDays = await knex('crop_stages')
+      .sum('duration as duration')
+      .where('fk_crops_id', batch[0].fk_crop_id);
```
Please update fk_crops_id to fk_crop_id in your knex migration and controller if it's being used because fk_crops_id has been updated to fk_crop_id to make it consistent and to avoid confusion on migrations schema and DB diagram. So please replace fk_crops_id to fk_crop_id if your application is crashing and you want to save time.

moonflare commented on src/server/api/controllers/status-box.controller.js:
```diff
@@ -0,0 +1,142 @@
+      throw new Error(`incorrect entry with the id of ${batchId}`, 404);
+    }
+
+    const cropStagedId = [1, 2, 3];
```
Please don't use hardcoded data. 🤔

moonflare commented on src/server/api/controllers/days-to-end-batch.controllers.js:
```diff
@@ -0,0 +1,31 @@
+    const end = moment();
+    const daysPassed = end.diff(start, 'days');
+    const daysLeft = totalDays[0].duration - daysPassed;
+    return `${daysLeft} days`;
```
I would return just the integer and in the frontend we add the `days` part. Also, you need to consider `1 day` case as well.

moonflare commented on src/server/api/controllers/days-to-end-batch.controllers.js:
```diff
@@ -0,0 +1,31 @@
+
+const getDayLeftToEndBatch = async (batchId) => {
+  try {
+    const batch = await knex('batches')
```
To avoid having `batch[0]` everywhere, we can have this, which means `const batch = result[0]`, or taking the first element of the result.
```suggestion
    const [batch] = await knex('batches')
```

cecastosic commented:
@moonflare Thanks for your review. We refactored this PR based on your feedback, can you please review it again. 
@sowmya1408 Thanks for your review. We added the middleware, can you please review it again? 
@mahieakhtar fk_crops_id is changed to fk_crop_id - but be aware endpoints will not work, untill PR #180 is not merged to develop

mahieakhtar commented:
> @moonflare Thanks for your review. We refactored this PR based on your feedback, can you please review it again.
> @sowmya1408 Thanks for your review. We added the middleware, can you please review it again?
> @mahieakhtar fk_crops_id is changed to fk_crop_id - but be aware endpoints will not work, untill PR #180 is not merged to develop

@Daniel Its time to review and merge PR#180, Thank you.

moonflare commented:
> > @moonflare Thanks for your review. We refactored this PR based on your feedback, can you please review it again.
> > @sowmya1408 Thanks for your review. We added the middleware, can you please review it again?
> > @mahieakhtar fk_crops_id is changed to fk_crop_id - but be aware endpoints will not work, untill PR #180 is not merged to develop
> 
> @daniel Its time to review and merge PR#180, Thank you.

I can have a look at #180 later tonight.

moonflare commented on src/server/config/swagger.json:
```diff
@@ -9,6 +9,98 @@
   "host": "",
   "basePath": "/api",
   "paths": {
+    "/batch-default-values/{batchId}": {
```
Did we add this route? or is it legacy code? I think we can remove it.

cecastosic replied:
This is another endpoint :)

moonflare replied:
Ah. I see it [here](https://github.com/HackYourFuture-CPH/FP-class11/blob/348853861cc9192bdacb39d89ae536be820f7f50/src/server/api/routes/api-router.js#L35). It was just added to swagger in this PR. My bad 😓

cecastosic commented:
> Nice work overall. But there are some performance improvements we should do before we merge this one. There is a lot of over fetching data. We fire the same queries over and over. If we call one query once, we already have that data, no need to call it again. For example the query below, we run it 3 times I think.
> 
> ```js
> knex('batches')
>       .select('seeding_date', 'fk_crop_id')
>       .where('id', batchId);
> ```
> 
> We should get all the data we need first, in one go if it makes sense, and then start calculating different metrics with that data. 🙌

Done :)

moonflare commented on src/server/api/controllers/batch-status.controller.js:
```diff
@@ -0,0 +1,152 @@
+const getBatchStatus = async (batchId) => {
+  try {
+    const batchesStages = await getBatchesStages(batchId);
+    const batch = await getBatch(batchId);
```
Considering that we only  get the `seeding_date` from the batchs table, I think we can extract it from the `batchesStages` like so `batchesStages[0].seeding_date` as we already have that data. This way we remove one extra request to the database 👌

moonflare replied:
Or perhaps remove `seeding_date` from `batchesStages` and only use the `seeding_date` provided by the `getBatch()` and this way we don't have the same `seeding_date` for each stage item inside `batchesStages` list. Up to you which one you pick 🙌

cecastosic replied:
Please check now 🙂

moonflare commented on src/server/api/controllers/batch-status.controller.js:
```diff
@@ -0,0 +1,152 @@
+  if (currentStage.length === 0) {
+    throw new Error(`incorrect entry with the id of ${batchId}`, 404);
+  }
+  if (currentStageName === 'delivered') return { status: currentStageName };
```
We should keep the data structure the same regardless if we have data or not. This way we only check for the values on the frontend.
```suggestion
  if (currentStageName === 'delivered') return { status: currentStageName, day: null };
```

moonflare commented on src/server/api/controllers/batch-status.controller.js:
```diff
@@ -0,0 +1,152 @@
+      .select('id')
+      .where('name', 'harvest');
+    const durationDaysBeforeHarvest = batchesStages
+      .filter((stage) => stage.id < harvestId.id)
```
I'm not sure I understand this filter. Can you please help me understand what is it for?

cecastosic replied:
Filtering all stages before harvest and summing up duration of those stages

moonflare commented on src/server/api/controllers/batch-status.controller.js:
```diff
@@ -0,0 +1,152 @@
+      moment(todayDate).format('DD-MM-YYYY') >=
+      moment(harvestStartDate).format('DD-MM-YYYY')
+    )
+      return `Harvest started`;
```
We should avoid returning this kind of data format. If our function returns integers when data is available, we should probably return `null` when there are no days and in the frontend we know if the result is `null`, then the `Harvest started`

dpfernandes commented:
Approved
